pep-9999.rst

```diff
@@ -0,0 +1,937 @@
+immutable object to reference a mutable object. This model is both
+easy to explain and understand, and an object's immutability can
+be "trusted" --- it is not possible for an immutable object to
+change through some nested mutable state [#RC]_. At the same time
+it limits the utility of freezing as many Python objects contain
+types outside of the standard library defined in C, which must
+opt-in immutability before they can be frozen.
+
+This PEP proposes immutability to be strict.
+
+
+Dealing with Failure During Freezing
+------------------------------------
+
+Regardless whether support for freezing is opt-in or opt-out some
+types will not be freezable. (Example such types include IO types
+like file handles, and caches -- as opposed to the cached
+objects.) This raises the question how to handle failure to freeze
+an object graph. Consider the object graph ``o1 --> o2 --> o3``
+where ``o1`` and ``o3`` can be frozen, but ``o2`` cannot. What are
+the possible behaviours of ``freeze(o1)``?
+
+1. Freeze fails partially. All subgraphs which could be frozen
+   entirely remain frozen. Remaining objects remain mutable. In
+   our example, ``o3`` remains frozen but ``o1`` and ``o2`` remain
+   mutable. This preserves strict immutability. The exception
+   thrown by the failing ``freeze(o1)`` call will contain ``o2``
+   (the place that caused freezing to fail) and ``o1`` (the object
+   in the graph that holds on to the failing object) to facilitate
+   debugging.
+2. **Rejected alternative**: Freeze fails completely. In the strict
+   interpretation of deep immutability, freezing ``o1`` is not
+   possible because ``o1`` contains a reference to an un-freezable
+   object ``o2``. In this scenario, the object graph ``o1 --> o2
+   --> o3`` remains mutable and ``freeze(o1)`` raises an exception
+   when the object graph traversal encounters ``o2``.
+3. **Rejected alternative**: Freeze succeeds by altering the
+   graph. In this example removing ``o2`` from the graph or
+   swapping out ``o2`` for a placeholder object to be able to
+   freeze the graph. This alternative becomes complicated both to
+   reason about from a user's perspective, and to implement when
+   ``o2`` is referenced multiple times.
+4. **Rejected alternative**: Permit the user to choose between
+   alternatives 1) and 3) at use-site. In this case, the
+   ``freeze`` function takes an optional 2nd argument ``strict``
+   which must either be ``True`` or ``False``. In the first case,
+   ``freeze`` behaves as in alternative 1), in the second case,
+   it behaves as in alternative 2). We could further track whether
+   an object is strictly immutable or not in order to prevent
+   non-strictly immutable objects to participate in operations
+   which require strictness. This adds additional complexity to
+   the implementation, and also for the user.
+   
+This PEP proposes following alternative 1, where freezing either
+succeeds or fails completely.
```

Comment:
Shouldn't this say: "or fails partially"? Option 2 would be failing completely.

```suggestion
This PEP proposes following alternative 1, where freezing either
succeeds or fails partially.
```

Or it could be that the number is wrong and this pep actually proposes success or failing completely which would be "2. rejected alternative"

---

I also wonder if we should add a motivation like "[...] fails partially. We believe this is easier to implement and more in the spirit of what the user wants."

But this could, of course, spark discussions